<<<<<<< HEAD
---
description: 'Disallow aliasing `this`.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-this-alias** for documentation.
=======
# Disallow aliasing `this` (`no-this-alias`)

This rule prohibits assigning variables to `this`.

## Rule Details

Rationale from TSLint:

> Assigning a variable to `this` instead of properly using arrow lambdas may be a symptom of pre-ES6 practices
> or not managing scope well.
>
> Instead of storing a reference to `this` and using it inside a `function () {`:
>
> ```js
> const self = this;
>
> setTimeout(function () {
>   self.doWork();
> });
> ```
>
> Use `() =>` arrow lambdas, as they preserve `this` scope for you:
>
> ```js
> setTimeout(() => {
>   this.doWork();
> });
> ```
>>>>>>> Render-FrontEnd

Assigning a variable to `this` instead of properly using arrow lambdas may be a symptom of pre-ES6 practices
or not managing scope well.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```js
const self = this;

<<<<<<< HEAD
setTimeout(function () {
  self.doWork();
});
=======
You can pass an object option:

```jsonc
{
  "@typescript-eslint/no-this-alias": [
    "error",
    {
      "allowDestructuring": true, // Allow `const { props, state } = this`; false by default
      "allowedNames": ["self"] // Allow `const self = this`; `[]` by default
    }
  ]
}
>>>>>>> Render-FrontEnd
```

### ‚úÖ Correct

```js
setTimeout(() => {
  this.doWork();
});
```

## Options

## When Not To Use It

If you need to assign `this` to variables, you shouldn‚Äôt use this rule.
